### ISSUE TITLE
Object.assign overloads don't allow generic target

### ISSUE THREAD

**Issue by turadg** [microsoft/TypeScript#48886]
# Bug Report

A function that uses generics to specify the type of the `target` parameter to `Object.assign` now fails.

### 🔎 Search Terms

- Object.assign
- target

### 🕗 Version & Regression Information

4.7 beta and nightly

- This changed between versions 4.6 and 47

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.7.0-dev.20220429#code/GYVwdgxgLglg9mABABwIYCdaoDYEEDO+MA5mADwAqAfIgBRQbECmUAXIhQDSL5wjoQm7AAoYs2SlQCUiAN4BYAFCJEAeQBGAKybQAdKkIkw9Ri269+gqQG4lAX1uKloSLAQoxMHASKkA6nDoANYYfGAAJpJ0DOjMbBzmfAJCiKKYXhLUMgrKalo6UPqGpLSydtwxcYmWTDb2jkA)

### 💻 Code

```ts
function partialAssign<T> (target: T, source: Partial<T>) {
  // this errors that T is not assignable to object or {}
  Object.assign(target, source);
};

function partialAssignWorkaround<T> (target: T, source: Partial<T>) {
  Object.assign({}, target, source);
};
```

### 🙁 Actual behavior

```
No overload matches this call.
  Overload 1 of 4, '(target: {}, source: Partial<T>): Partial<T>', gave the following error.
    Argument of type 'T' is not assignable to parameter of type '{}'.
  Overload 2 of 4, '(target: object, ...sources: any[]): any', gave the following error.
    Argument of type 'T' is not assignable to parameter of type 'object'.
```

### 🙂 Expected behavior

No error.

**Comment by Josh-Cena** [comment]
This is expected. The generic may be instantiated as `null`, and `partialAssign(null, [anything])` would give an error at runtime. This bug was fixed in 4.7: https://devblogs.microsoft.com/typescript/announcing-typescript-4-7-beta/#type-parameters-no-longer-assignable-to-in-strictnullchecks Simply adding a `<T extends object>` would fix the issue.

**Comment by turadg** [comment]
Indeed. Thanks for the reference and solution.
